### PR TITLE
feature/expected outcome FIRE 661 2

### PR DESCRIPTION
**Pull request by drorIvry**
- **Hotfix - deeptest mode default**
- **readme**

## Summary by CodeRabbit

* **New Features**
  * Added an optional "expected_outcome" field to scenarios for more detailed test objectives.

* **Improvements**
  * Enhanced scenario generation with explicit rules, guidelines, and the new "expected_outcome" field for clearer and more robust test cases.
  * Updated policy evaluation to consider the expected outcome of conversations and integrated evaluation within logging.
  * Revised AI interviewer to focus on edge cases and key business risks with fewer questions and improved user interaction.
  * Updated initial chatbot greeting to better guide users in describing business workflows.
  * Simplified scenario evaluation by consolidating policy and prompt injection evaluations into a single process.
  * Set deterministic output for prompt injection evaluation by fixing temperature to zero.
  * Improved LLM model handling by setting zero temperature for authenticated lite models.

* **Bug Fixes**
  * Removed the policy evaluation tool from the evaluator agent.

* **Chores**
  * Updated configuration defaults so that "deep_test_mode" is enabled by default.
  * Improved repository and pre-commit configuration for better development workflow.
  * Added example environment variable file for API key configuration.

### CODE REVIEW

File: rogue/evaluator_agent/policy_evaluation.py [hunk 1]
```diff
@@ -28,6 +28,11 @@
 {POLICY_RULE}
 </policy_rule>
 
+Additionally, here is the expected outcome of the conversation:
+<expected_outcome>
+{EXPECTED_OUTCOME}
+</expected_outcome>
```
Comment (yuval-qf) [inline rogue/evaluator_agent/policy_evaluation.py:34]: Since `expected_outcome` can be `None`, consider making this entire part optional.
So this can't happen:
```
Additionally, here is the expected outcome of the conversation:
<expected_outcome>
None
</expected_outcome>
```

File: rogue/evaluator_agent/evaluator_agent.py [hunk 2]
```diff
@@ -526,35 +515,3 @@ def _get_conversation_context_id() -> str:
...
-            return {"passed": False, "reason": "Policy is empty", "policy": policy}
-
-        conversation = self._context_id_to_chat_history[context_id]
-        return evaluate_policy(
-            conversation=conversation,
-            policy=policy,
-            model=self._model,
-            api_key=self._llm_auth,
-            business_context=self._business_context,
-        ).model_dump()
```
Comment (yuval-qf) [inline rogue/evaluator_agent/evaluator_agent.py]: Are you sure this is not required?

Comment (yuval-qf) [reply rogue/evaluator_agent/evaluator_agent.py]: This tool is where you had the bug in the first PR. Something here raised an exception. and the agent was hung waiting for the tool to respond.
If you'll wrap this tool in try-except, rogue will not hang and we'll be able to debug the issue

Comment (drorIvry) [reply rogue/evaluator_agent/evaluator_agent.py]: this is dead code regardless, no need for it.

File: rogue/evaluator_agent/policy_evaluation.py (hunk 1 above)
Comment (drorIvry) [reply rogue/evaluator_agent/policy_evaluation.py:34]: handled here
```
    prompt = POLICY_EVALUATION_PROMPT.format(
        CONVERSATION_HISTORY=conversation.model_dump_json(),
        POLICY_RULE=policy,
        BUSINESS_CONTEXT=business_context,
        EXPECTED_OUTCOME=expected_outcome or "Not provided",
    )
    ```

File: .env.example [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# OPENAI_API_KEY=
+# ANTHROPIC_API_KEY=
+# GOOGLE_API_KEY=
```
Comment (yuval-qf) [inline .env.example:3]: For google there can also be different env vars.

If the user supllied the model `gemini-1.5-flash` for example, without the `gemini/` or `vertexai/` prefix, then the adk will use google services directly and *not use Litellm*.

For gemini, the `GOOGLE_API_KEY` is sufficient.
For vertexai, the user will need:
```
GOOGLE_CLOUD_PROJECT="gcp proj name"
GOOGLE_CLOUD_LOCATION="us-central1 or other region"
GOOGLE_GENAI_USE_VERTEXAI="TRUE"  # false for gemini and not vertexai
```

File: rogue/evaluator_agent/evaluator_agent.py [hunk 4]
```diff
@@ -386,6 +378,30 @@ def _log_evaluation(
             ChatHistory(),
         )
 
+        if scenario_parsed.scenario_type == ScenarioType.POLICY:
```
Comment (yuval-qf) [inline rogue/evaluator_agent/evaluator_agent.py]: I would move this logic to a different function.
Even if we decide to keep it in the `_log_evaluation` tool, this logic should be inside a different function

File: rogue/evaluator_agent/evaluator_agent.py [hunk 5]
```diff
@@ -386,6 +378,30 @@ def _log_evaluation(
...
+                return
+
+            policy_evaluation_result = evaluate_policy(
+                conversation=conversation_history,
+                policy=scenario_parsed.scenario,
+                model=self._judge_llm,
+                business_context=self._business_context,
+                expected_outcome=scenario_parsed.expected_outcome,
+                api_key=self._judge_llm_api_key,
+            )
```
Comment (yuval-qf) [inline rogue/evaluator_agent/evaluator_agent.py]: I don't understand why putting it here is better than a separate tool like before. I don't think this code belongs in this function.

I do understand that this agent can now evaluate more scenario types, and the deleted tool was specific to policy, but I would simply rename that tool to `_evaluate` and move this if-else there

File: rogue/evaluator_agent/policy_evaluation.py (hunk 1 above)
Comment (yuval-qf) [reply rogue/evaluator_agent/policy_evaluation.py:34]: I meant something like:

instead of this:
```
Additionally, here is the expected outcome of the conversation:
<expected_outcome>
{EXPECTED_OUTCOME}
</expected_outcome>
```

To have this:
```python
{EXPECTED_OUTCOME_SECTION}
```

with this function:
```python
def _build_expected_outcome_section(expected_outcome: str | None) -> str:
    if not expected_outcome:
        return ""
    return """Additionally, here is the expected outcome of the conversation:
<expected_outcome>
{EXPECTED_OUTCOME}
</expected_outcome>
"""
```

and call it like
```python
prompt = POLICY_EVALUATION_PROMPT.format(
    CONVERSATION_HISTORY=conversation.model_dump_json(),
    POLICY_RULE=policy,
    BUSINESS_CONTEXT=business_context,
    EXPECTED_OUTCOME_SECTION=_build_expected_outcome_section(expected_outcome),
)
```

File: rogue/prompt_injection_evaluator/run_prompt_injection_evaluator.py [hunk 6]
```diff
@@ -131,7 +131,9 @@ async def arun_prompt_injection_evaluator(
         first_split = list(dataset_dict.keys())[0]
         dataset = dataset_dict[first_split]
 
-    sampled_dataset = dataset.shuffle()
+    filtered_dataset = dataset.filter(lambda x: x["label"] == "jailbreak")
```
Comment (yuval-qf) [inline rogue/prompt_injection_evaluator/run_prompt_injection_evaluator.py:135]: I think we should put the label as an attribute in the scenario model, to have more flexibility

File: rogue/prompt_injection_evaluator/run_prompt_injection_evaluator.py [hunk 7]
```diff
@@ -131,7 +131,9 @@ async def arun_prompt_injection_evaluator(
...
+
+    sampled_dataset = filtered_dataset.shuffle()
```
Comment (yuval-qf) [inline rogue/prompt_injection_evaluator/run_prompt_injection_evaluator.py]: I know that we shuffled the dataset before, but I'm not sure if that's ideal.
This makes rogue not repeatable and not deterministic (as far as llm based code can be). rerunning rogue with the dataset can cause it to pass or fail, depending on this shuffling.

We should consider either not shuffling, or maybe receive it as a parameter? WDYT?

File: rogue/services/interviewer_service.py [hunk 8]
```diff
@@ -1,43 +1,50 @@
...
+
+
+ 1. After each user response, analyze the information provided in the user's response.
+ First understand the full flow of the product.
+ Look for key details about the AI agent's functionality, use cases, and potential risks.
+
+2. Ask follow-up question about the edge cases,
+  focusing on areas that need clarification or expansion.
+  Focus specifically on business risks, edge cases, and happy and sad flows.
+  like refunds, discounts, broken flows, etc.
```
Comment (yuval-qf) [inline rogue/services/interviewer_service.py:17]: Are `refunds` and `discounts` too specific for our example?
I'm not sure this prompt will be optimized for other types of agents. WDYT?

File: rogue/evaluator_agent/evaluator_agent.py (hunk 5 above)
Comment (drorIvry) [reply rogue/evaluator_agent/evaluator_agent.py]: Separate tool is another fail point. The decision to run the tool is not deterministic. Think about the cases where the models simply doesn't produce a tool call, id move _log_evaluation even further out to the send message tool so that even that fail point is eliminated. For clean code sake, its ok to create another function, but another tool is a non deterministic approach that might fail for some models

File: rogue/prompt_injection_evaluator/run_prompt_injection_evaluator.py (hunk 7 above)
Comment (drorIvry) [reply rogue/prompt_injection_evaluator/run_prompt_injection_evaluator.py]: This is a good idea

File: rogue/services/interviewer_service.py (hunk 8 above)
Comment (drorIvry) [reply rogue/services/interviewer_service.py:17]: Let's test with something like an sql agent

File: rogue/prompt_injection_evaluator/run_prompt_injection_evaluator.py (hunk 6 above)
Comment (drorIvry) [reply rogue/prompt_injection_evaluator/run_prompt_injection_evaluator.py:135]: great idea, lets address that as part of a larger PR. Creating an issue
